Handle resilient stored properties in objcImpl

beccadax opened the pull request:
This PR temporarily bans, and lays the groundwork for eventually supporting, resilient stored properties in `@objc @implementation` classes.

When a stored property contains a resilient value type, its size cannot be determined at compile time. That means the size and exact layout of the instance variables is not known, either. The layout must be computed, and the instance variables and class size data updated, when the class is realized (i.e. at runtime, but before it is used).

For Swift classes (including ordinary `@objc` classes), the mechanism used to do this works like so:

1. When the Objective-C runtime realizes a class, it checks a few flags and notices that the class has an "ObjC metadata update function", which it invokes.
2. The ObjC metadata update function retrieves the Swift type metadata for the class.
3. If the Swift type metadata is not yet complete, this will cause a "Swift type metadata completion function" to be run.
4. The Swift type metadata completion function gathers information about the sizes of the types in the stored properties (which are now known since it's runtime) and calls the `swift_updateClassMetadata2()` function to update the instance variable layout and tell the ObjC runtime to finish realizing the class.

This happens before the Objective-C runtime does its own ivar sliding. In effect, the ivars are slid twice—once to account for changes in the sizes of other ivars ahead of them in the class layout, then again to account for changes in the size of the superclass.

Unfortunately, every single step in this chain is designed for full-fledged Swift classes with Swift metadata; all of them break horribly if you try to use them with a "pure ObjC" class like the ones created by `@objc @implementation`:

1. The Objective-C runtime doesn't check for or call the ObjC metadata update function unless the "Swift bit" is set on the class. The Swift bit is normally used to indicate the presence of Swift metadata, so it's not supposed to be set on pure ObjC classes.
2. Retrieving the Swift type metadata requires use of the type metadata symbol, which is not present for pure ObjC classes.
3. The Swift type metadata completion function is specified in Swift metadata and there's no way to attach one to a pure ObjC class.
4. The `swift_updateClassMetadata2()` function assumes it has been passed a pointer to Swift class metadata and both reads and writes fields that are not present in pure ObjC classes.

This PR lays the groundwork for an alternate mechanism to update the layout of a pure ObjC class during class realization:

* The compiler will generate a different ObjC metadata update function which gathers type size information and directly passes it to the runtime, rather than using Swift metadata access mechanisms. This sidesteps problems 2 and 3.
* It will call a new runtime entry point, `swift_updatePureObjCClassMetadata()`, which uses only the Objective-C metadata. This sidesteps problem 4.

However, problem 1 will require changes to the Objective-C runtime, so the new code generation is gated behind an experimental feature flag, `ObjCImplementationWithResilientStorage`. Without that flag, Swift will unconditionally reject properties that would require this new mechanism; with it, Swift will enforce a (for now) 99.99 minimum deployment target for it.

Fixes rdar://126948464.

beccadax committed:
[NFC] Refactor objcImpl matadata layout logic

Rather than having individual methods return without doing anything, arrange for ClassMetadataVisitor to never call them in the first place. This makes it so ClassMetadataScanner also knows which fields are omitted and it can compute offsets correctly.

The old logic is still present for field offsets to make this change NFC.

beccadax committed:
[NFC] Refactor swift_updatePureObjCClassMetadata()

Merge the three-stage operation originally designed for field vectors into a single unified loop that acts directly on the ivar offsets instead of using a faux field offset vector.

beccadax committed:
Fix objc_implementation_objc_client test

Because this test is disabled, a previous commit failed to add an experimental feature flag that’s required to run it.

beccadax committed:
Hide objcImpl resilience support behind feature

Resilence support will require changes to the Objective-C runtime to expand support for metadata initialization functions. Add a separate experimental feature flag to help with staging that support in, and modify diagnostics to not suggest increasing the minimum deployment target for now.

beccadax committed:
Set a minimum deployment target for objcImpl

We’re not committing to @objc @implementation back-deploying to pre-stable Apple platforms.

beccadax committed:
Diagnose resilient properties in objcImpl

…when the deployment target is not high enough to support them.

beccadax commented:
@swift-ci please test

beccadax committed:
Add target to failing objcImpl test

beccadax commented:
@swift-ci please test

beccadax committed:
Fix -target in several objcImpl tests

beccadax commented:
@swift-ci please test

beccadax commented:
@swift-ci please test

tshortli commented on lib/IRGen/GenMeta.cpp:
```diff
@@ -2948,37 +2948,40 @@ IRGenModule::getAddrOfOriginalModuleContextDescriptor(StringRef Name) {
 
   // Fill out an array with the field type metadata records.
-  Address fields = IGF.createAlloca(
+  Address fields = createAlloca(
```
Minor: it looks like some code in this file hasn't been reformatted after changes

beccadax replied:
Yeah, I initially wasn't sure if this change would stick. I should go over this again.

tshortli commented on lib/IRGen/GenMeta.cpp:
```diff
@@ -5054,13 +5057,88 @@ static void emitObjCClassSymbol(IRGenModule &IGM, ClassDecl *classDecl,
+    // Update and FixedOrUpdate require support in both the Swift and ObjC
+    // runtimes.
+    auto requiredAvailability=ctx.getUpdatePureObjCClassMetadataAvailability();
+    // FIXME: Take the class's availability into account
```
Out of curiosity, does class metadata layout happen lazily when the first instance is created or something? I'm wondering how gating this on the availability of the class itself would be safe in practice.

beccadax replied:
The first time the ObjC runtime tries to access any of the metadata associated with a class—basically, use it as anything *but* a completely opaque pointer—it performs an operation called "realization". The metadata update function that Swift generates will be called during realization, and it will call the new `swift_updatePureObjCClassMetadata()` function. So realization might happen when you try to look up the `+alloc` method to allocate an instance of the class, or it could happen even earlier, depending on how the code uses the class.

tshortli commented on test/IRGen/objc_implementation.swift:
```diff
@@ -1,34 +1,34 @@
 // REQUIRES: OS=macosx
 
-// RUN: %target-swift-frontend(mock-sdk: %clang-importer-sdk) -enable-experimental-feature CImplementation -I %S/Inputs/abi -F %clang-importer-sdk-path/frameworks %s -import-objc-header %S/Inputs/objc_implementation.h -emit-ir > %t.ir
+// RUN: %target-swift-frontend(mock-sdk: %clang-importer-sdk) -enable-experimental-feature CImplementation -I %S/Inputs/abi -F %clang-importer-sdk-path/frameworks %s -import-objc-header %S/Inputs/objc_implementation.h -emit-ir -target %target-future-triple > %t.ir
```
Not sure if it would make sense here, but we now have lit substitutions like `%target-swift-abi-6.0-triple` if you wanted to specify a specific, supported deployment target.

beccadax committed:
Eliminate objcImpl field offset vectors

We really don’t need ‘em; we can just adjust the direct field offsets.

The runtime entry point currently uses a weird little hack that we will refactor away shortly.

beccadax committed:
Support resilient stored properties in objcImpl

When an @objc @implementation class requires the use of `ClassMetadataStrategy::Update` because some of its stored properties do not have fixed sizes, we adjust the direct field offsets during class realization by emitting a custom metadata update function which calls a new entry point in the Swift runtime. That entry point adjusts field offsets like `swift_updateClassMetadata2()`, but it only assumes that the class has Objective-C metadata, not Swift metadata.

This commit introduces an alternative mechanism which does the same thing without using any Swift-only metadata. It’s a rough implementation with important limitations:

• We’re currently using the field offset vector, which means that field offsets are being emitted into @objc @implementation classes; these will be removed.
• The new Swift runtime entry point duplicates a lot of `swift_updateClassMetadata2()`’s implementation; it will be refactored into something much smaller and more compact.
• Availability bounds for this feature have not yet been implemented.

Future commits in this PR will correct these issues.

beccadax commented:
Made some formatting improvements at @tshortli's request. No functional changes.

beccadax commented:
@swift-ci please smoke test

beccadax commented:
Windows build timed out.

beccadax commented:
@swift-ci please smoke test Windows platform
